[acc][flang][cir] Add recipes to data entry operations

razvanlupusoru opened the pull request:
This patch refactors the OpenACC dialect to attach recipe symbols
directly to data operations (acc.private, acc.firstprivate, acc.reduction)
rather than to compute constructs (acc.parallel, acc.serial, acc.loop).

Motivation:
The previous design required compute constructs to carry both the recipe
symbol and the variable reference, leading to complexity. Additionally,
recipes were required even when they could be generated automatically
through MappableType interfaces.

Changes:
- Data operations (acc.private, acc.firstprivate, acc.reduction) now
  require a 'recipe' attribute referencing their respective recipe
  operations
- Verifier enforces recipe attribute presence for non-MappableType
  operands; MappableType operands can generate recipes on demand
- Compute constructs (acc.parallel, acc.serial, acc.loop) no longer
  carry recipe symbols in their operands
- Updated flang lowering to attach recipes to data operations instead
  of passing them to compute constructs

Format Migration:
Old format:
```
  acc.parallel private(@recipe -> %var : !fir.ref<i32>) { ... }
```

New format:
```
  %private = acc.private varPtr(%var : !fir.ref<i32>)
              recipe(@recipe) -> !fir.ref<i32>
  acc.parallel private(%private : !fir.ref<i32>) { ... }
```

Test Updates:
- Updated all CIR and Flang OpenACC tests to new format
- Fixed CHECK lines to verify recipe attributes on data operations

Commit:
[draft][acc][flang] Add recipes to data entry operations

Currently recipes are held as part of the constructs. But
the recipes are needed specifically for materializing semantics
of the data clauses. There is no reason to not have this on the
data clauses themselves.

TODO: Fix the IR in flang tests which do not expect new operand
TODO: Deprecate the recipe list held on the constructs

Commit:
Merge remote-tracking branch 'upstream/main' into accdatarecipe

Commit:
Remove construct recipes and fix all tests

Commit:
Update invalid test to include new messages

Commit:
Fix format

Commit:
Restore the data operands verifier

Commit:
Add missed invalid cases

erichkeane reviewed (commented):
I didn't quite get all the patch, but I was wondering if this is going to reject the old mechanism?  It would be great for me if it did, as that'll make sure I get it right in the clang implementation after rebase :)

Commit:
Merge remote-tracking branch 'upstream/main' into accdatarecipe

Commit:
Merge remote-tracking branch 'upstream/main' into accdatarecipe

Commit:
Fix ops.mlir test

Commit:
Fix ACCImplicitData and tests

Commit:
Fix flang ACCRecipeBufferization

Commit:
Fix flang tests

Commit:
Fix CIR tests

Commit:
Fix more flang tests

Commit:
Fix format

Commit:
Remove .str()

razvanlupusoru commented:
I have rebased this PR. And then fixed all of the passes and tests that were added after I did the initial work in July. Finally, I also fixed all of the CIR tests also.

vzakhari commented on flang/lib/Lower/OpenACC.cpp:
```diff
@@ -5142,6 +5106,7 @@ mlir::Operation *Fortran::lower::genOpenACCLoopFromDoConstruct(
   // So this means that in all cases we mark with `auto` unless it is a
   // `do concurrent` in an `acc parallel` construct or it must be `seq` because
   // it is in an `acc serial` construct.
+  fir::FirOpBuilder &builder = converter.getFirOpBuilder();
```
Unused variable?

vzakhari reviewed (commented):
Flang and MLIR/OpenACC changes look great to me! Thank you, Razvan!

clementval reviewed (approved):
LGTM

erichkeane reviewed (approved):
This is perfectly sensible, +1.  I did a spot check of all the CIR tests and every one I looked at seemed perfect, and this appears pretty formulaic, so LGTM!

Commit:
Do not require recipe for MappableType

Commit:
Merge remote-tracking branch 'upstream/main' into accdatarecipe

Commit:
Reuse builder

Commit:
Merge remote-tracking branch 'upstream/main' into accdatarecipe

Commit:
Fix newly added acc-reduction-remapping.f90 test
